src/valid/mod.rs

```diff
@@ -0,0 +1,382 @@
+mod analyzer;
+mod expression;
+mod function;
+mod interface;
+
+use crate::{
+    arena::{Arena, Handle},
+    proc::Typifier,
+    FastHashSet,
+};
+use bit_set::BitSet;
+use thiserror::Error;
+
+//TODO: analyze the model at the same time as we validate it,
+// merge the corresponding matches over expressions and statements.
```

Comment:
So to be clear: this is now done in two passes? First analyzing then validating?

Reply:
correct. The PR doesn't change this part of the logic, it just re-groups it closer so that we can merge it some time later